src/WindowsUtils/DllInjection.cpp

```diff
@@ -0,0 +1,278 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "WindowsUtils/DllInjection.h"
+
+#include <absl/base/casts.h>
+#include <absl/strings/match.h>
+#include <absl/strings/str_format.h>
+
+#include "OrbitBase/GetLastError.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/UniqueResource.h"
+#include "WindowsUtils/ListModules.h"
+
+// clang-format off
+#include <windows.h>
+#include <processthreadsapi.h>
+// clang-format on
+
+namespace orbit_windows_utils {
+
+namespace {
+
+ErrorMessageOr<uint64_t> RemoteWrite(HANDLE process_handle, const std::vector<uint8_t>& buffer) {
+  // Allocate memory in target process.
+  LPVOID base_address = VirtualAllocEx(process_handle, /*lpAddress=*/0, buffer.size(),
+                                       MEM_RESERVE | MEM_COMMIT, PAGE_READWRITE);
+  if (base_address == nullptr) {
+    return ErrorMessage(
+        absl::StrFormat("VirtualAllocEx failed with error %s", orbit_base::GetLastErrorAsString()));
+  }
+
+  // Write in allocated remote memory.
+  size_t num_bytes_written = 0;
+  if (!WriteProcessMemory(process_handle, base_address, static_cast<LPCVOID>(buffer.data()),
+                          buffer.size(), &num_bytes_written)) {
+    return ErrorMessage(absl::StrFormat("WriteProcessMemory failed with error %s",
+                                        orbit_base::GetLastErrorAsString()));
+  }
+
+  if (num_bytes_written != buffer.size()) {
+    return ErrorMessage("WriteProcessMemory could not write the requested number bytes.");
+  }
+
+  return absl::bit_cast<uint64_t>(base_address);
+}
+
+template <typename T>
+ErrorMessageOr<void> RemoteRead(HANDLE process_handle, uint64_t base_address, T& result) {
```

Comment:
Why not return `ErrorMessageOr<T>`? Is it a lot of data and can the compiler not optimize? `ErrorMessageOr<std::string>` from `ReadRemoteString` seems cleaner.

Reply:
Yeah, I implemented your suggestion, it saved a few lines of code. The drawback is that we can't fill an existing object without a copy, but we don't have that use case here. Worse case we can add the previous implementation.